gpio/zynq7000-pwm/pwm.c

```diff
@@ -0,0 +1,243 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Xilinx Zynq 7000 PWM driver
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Aleksander Kaminski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include <sys/types.h>
+#include <sys/msg.h>
+#include <sys/threads.h>
+#include <posix/idtree.h>
+#include <posix/utils.h>
+#include <sys/mman.h>
+
+#define ROOTFS_WAIT 1
+#define PRIORITY    2
+
+#define RELOAD_DEFAULT 100000
+#define PWM_NCHAN      8
+#define PWM_RELOAD     PWM_NCHAN
+
+
+typedef struct {
+	oid_t oid;
+	int reg;
+} pwm_chan_t;
+
+
+static struct {
+	volatile uint32_t *base;
+	pwm_chan_t channel[PWM_NCHAN];
+} common;
+
+
+static void pwm_read(pwm_chan_t *chan, uint32_t *val)
+{
+	*val = *(common.base + chan->reg);
+}
+
+
+static void pwm_write(pwm_chan_t *chan, uint32_t val)
+{
+	uint32_t reload = *(common.base + PWM_RELOAD);
+
+	if (reload < val) {
+		val = reload;
+	}
+
+	*(common.base + chan->reg) = val;
+}
+
+
+static void pwm_setReload(uint32_t reload)
+{
+	*(common.base + PWM_RELOAD) = reload;
+}
+
+
+static void pwm_thread(void *arg)
+{
+	msg_t msg;
+	unsigned long rid;
+	uint32_t val;
+	int ret;
+	char buff[16];
+
+	(void)arg;
+
+	for (;;) {
+		if (msgRecv(common.channel[0].oid.port, &msg, &rid) < 0) {
+			continue;
+		}
+
+		switch (msg.type) {
+			case mtOpen:
+			case mtClose:
+				if (msg.i.io.oid.id >= PWM_NCHAN) {
+					msg.o.io.err = -ENOENT;
+				}
+				else {
+					msg.o.io.err = 0;
+				}
+				break;
+
+			case mtRead:
+				if (msg.i.io.oid.id >= PWM_NCHAN) {
+					msg.o.io.err = -ENOENT;
+				}
+				else {
+					pwm_read(&common.channel[msg.i.io.oid.id], &val);
+					ret = sprintf(buff, "%u\n", val);
+
+					if (ret < msg.i.io.offs) {
+						/* EOF */
+						msg.o.io.err = 0;
+					}
+					else {
+						strncpy(msg.o.data, buff + msg.i.io.offs, msg.o.size);
+						((char *)msg.o.data)[msg.o.size - 1] = '\0';
+						msg.o.io.err = ((ret + 1) < (int)msg.o.size) ? ret + 1 : msg.o.size;
+					}
+				}
+				break;
+
+			case mtWrite:
+				if (msg.i.io.oid.id >= PWM_NCHAN) {
+					msg.o.io.err = -ENOENT;
+				}
+				else if (msg.i.data == NULL || msg.i.size == 0) {
+					msg.o.io.err = -EINVAL;
+				}
+				else {
+					strncpy(buff, msg.i.data, msg.i.size > sizeof(buff) ? msg.i.size : sizeof(buff));
```

Comment:
I got the condition wrong way (should be msg.i.size < sizeof(buff)), thanks., will fix that. I want to include msg.i.size because this is data received from outside - I have no guarantee that it's NULL terminated.

Reply:
That's the reason I copy it in the first place - to be able to add NULL terminator if missing (msg.i.data is not const, but should be and will be in the future).